crates/ra_hir_ty/src/expr.rs

```diff
@@ -61,6 +63,37 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
         }
     }
 
+    fn validate_match(
+        &mut self,
+        id: ExprId,
+        expr: ExprId,
+        arms: &[MatchArm],
+        db: &dyn HirDatabase,
+    ) {
+        let (body, source_map): (Arc<Body>, Arc<BodySourceMap>)
+                                 = db.body_with_source_map(self.func.into());
+
+        let match_expr: &hir_def::expr::Expr = &body[expr];
+
+        // TODO do validation
+        // should I be doing validation against the hir_def::expr::Expr?
```

Comment:
yes, these kinds of analyses shouldn't touch the AST at all. Actually, you shouldn't even be using the `body_with_source_map` query, but just `body` (probably just pass the body as an argument).

Reply:
Thanks for the feedback. My usage of `body_with_source_map` was inspired by the `validate_record_literal` method, which uses it to build up the `MissingField` diagnostic (which contains an `AstPtr`). Just to clarify, do you mean the analysis shouldn't use the AST, or even the diagnostic shouldn't use the AST?

Reply:
Well, this is about two concerns mostly: 1. Using the abstractions of the HIR for analysis because that's what it's for, and 2. incrementality -- calling the `body_with_source_map` means this has to be recalculated whenever the AST changes in any way (i.e. if I type a space in one function, this would need to be rerun for all following functions in the file). `validate_record_literal` is probably fine because it only calls `body_with_source_map` if there actually is a diagnostic, which should not be the case for most functions. So using the source map in the end to provide the AST pointer is probably fine, but it should only be done if there are diagnostics.

Reply:
> Using the abstractions of the HIR for analysis because that's what it's for

For example, consider macros in patterns. I'm actually not sure whether we expand them yet, but if we do, the HIR body will already contain the expanded form, so this code doesn't need to be concerned with them at all.

Reply:
I believe I am only using the AST in creating the diagnostic, to report the error. This is in line with other diagnostics as far as I can tell. 

Has your concern here been resolved?